### PR TITLE
[Web LA] Custom `Keyframe` animations

### PR DESCRIPTION
**Pull request by m-bert**
## Summary

This PR adds support for custom layout animations created via `Keyframe` on web.

> [!IMPORTANT]
> This PR replaces #5277 

## Test plan

Tested on example app and the following code snippet:

<details>
<summary> Test code </summary>

```jsx
import { StyleSheet, View, Text, Pressable } from 'react-native';

import React, { useState } from 'react';
import Animated, { Easing, Keyframe } from 'react-native-reanimated';

const entering = new Keyframe({
  0: {
    transform: [
      { translateX: -500 },
      { translateY: -300 },
      { scale: 1.25 },
      { skewY: '25deg' },
    ],
    opacity: 0.3,
    easing: Easing.cubic,
  },

  70: {
    transform: [{ translateX: 250 }, { translateY: 100 }, { scale: 1.25 }],
    opacity: 0.7,
  },
});
const exiting = new Keyframe({
  0: {
    easing: Easing.exp,
  },
  100: {
    transform: [
      { translateX: 700 },
      { translateY: 250 },
      { scale: 0.3 },
      { rotate: '225deg' },
    ],
    opacity: 0,
  },
});

export default function EmptyExample() {
  const [show, setShow] = useState(true);
  return (
    <View style={styles.container}>
      {show && (
        <Animated.View
          entering={entering.duration(1000)}
          exiting={exiting}
          style={styles.box}
        />
      )}

      <Pressable onPress={() => setShow((prev) => !prev)} style={styles.button}>
        <Text style={{ color: 'white' }}> Click me! </Text>
      </Pressable>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
  },
  box: {
    width: 250,
    height: 250,
    backgroundColor: '#782aeb',
  },
  button: {
    width: 100,
    height: 50,
    backgroundColor: '#57b495',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-around',
    borderRadius: 15,

    position: 'absolute',
    top: 10,
    left: 10,
  },
});

```

</details>

### CODE REVIEW

File: packages/react-native-reanimated/src/layoutReanimation/web/Easing.web.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+'use strict';
+
+// Those are the easings that can be implemented using Bezier curves.
+// Others should be done as CSS animations
+export const WebEasings = {
+  linear: [0, 0, 1, 1],
```
[the remark this answers is not in the document]
Comment (m-bert) [reply packages/react-native-reanimated/src/layoutReanimation/web/Easing.web.ts:6]: We have 2 sources:

- GPT
- Found empirically with @kacperkapusciak using reanimated docs.

Are you sure that we want to have that in code? 😅

File: packages/react-native-reanimated/src/layoutReanimation/web/animationsManager.ts [hunk 2]
```diff
@@ -111,25 +119,43 @@ function tryGetAnimationConfig<ComponentProps extends Record<string, unknown>>(
...
-      : (config.constructor as ConstructorWithStaticContext).presetName;
+  const isCustomKeyframe = config instanceof Keyframe;
+
+  const animationName = isCustomKeyframe
+    ? createCustomKeyFrameAnimation(
+        (config as CustomConfig).definitions as KeyframeDefinitions
+      )
+    : typeof config === 'function'
+    ? config.presetName
+    : (config.constructor as ConstructorWithStaticContext).presetName;
```
[the remark this answers is not in the document]
Comment (m-bert) [reply packages/react-native-reanimated/src/layoutReanimation/web/animationsManager.ts:130]: Sure, the only reason why I used `? :` was to keep this variable as `const`. However, I agree that it is not the easiest piece of code to read

[the remark this answers is not in the document]
Comment (m-bert) [reply packages/react-native-reanimated/src/layoutReanimation/web/animationsManager.ts]: Done in [55aab5e](https://github.com/software-mansion/react-native-reanimated/pull/6135/commits/55aab5e404ec89587b6ebcaa82852019e4998295)

File: packages/react-native-reanimated/src/layoutReanimation/web/componentUtils.ts [hunk 3]
```diff
@@ -63,12 +66,12 @@ export function getReducedMotionFromConfig(config: CustomConfig) {
...
-  animationName: AnimationNames
+  animationName: string
 ): number {
-  const defaultDuration = isLayoutTransition
-    ? 0.3
-    : Animations[animationName].duration;
+  const defaultDuration =
+    animationName in Animations
+      ? Animations[animationName as AnimationNames].duration
+      : 0.3;
```
[the remark this answers is not in the document]
Comment (m-bert) [reply packages/react-native-reanimated/src/layoutReanimation/web/componentUtils.ts:77]: Done in [49a9a58](https://github.com/software-mansion/react-native-reanimated/pull/6135/commits/49a9a58214d3b8e525b3589d2a792d4a120d55b6)

File: packages/react-native-reanimated/src/layoutReanimation/web/Easing.web.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (piaskowyk) [reply packages/react-native-reanimated/src/layoutReanimation/web/Easing.web.ts:6]: XD

[the remark this answers is not in the document]
Comment (piaskowyk) [reply packages/react-native-reanimated/src/layoutReanimation/web/Easing.web.ts:6]: so, current form is actually good enough 😅

**Review by piaskowyk — APPROVED** [review]

**Pull request merged**